cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,260 @@
+      }
+    }
+
+    // check if we have enough input samples
+    bool output_is_valid = (count >= min_periods);
+
+    // set the mask
+    MaskType const result_mask{__ballot_sync(active_threads, output_is_valid)};
+    gdf_index_type const out_mask_location = i / BITS_PER_MASK;
+    MaskType* const __restrict__ out_mask32 = reinterpret_cast<MaskType*>(out_col_valid);
+
+    // only one thread writes the mask
+    if (0 == threadIdx.x % warpSize)
+      out_mask32[out_mask_location] = result_mask;
+
+    // store the output value, one per thread
+    if (output_is_valid) {
+      if (average)
+        // special case for COUNT aggregator
+        out_col[i] = val / count;
+      else
+        out_col[i] = val;
+    }
+
+    // process next element 
+    i += stride;
+    active_threads = __ballot_sync(active_threads, i < nrows);
+  }
+}
+
+struct launch_kernel
+{
+  /**
+   * @brief Uses SFINAE to instantiate only for arithmetic types
+   */
+  template<typename ColumnType, template <typename AggType> class agg_op, bool average, class... TArgs,
+	   typename std::enable_if_t<std::is_arithmetic<ColumnType>::value, std::nullptr_t> = nullptr> 
+  void dispatch_aggregation_type(gdf_size_type nrows, TArgs... FArgs)
+  {
+    PUSH_RANGE("CUDF_ROLLING", GDF_ORANGE);
+
+    gdf_size_type block = 256;
+    gdf_size_type grid = (nrows + block-1) / block;
+
+    gpu_rolling<ColumnType, agg_op, average><<<grid, block>>>(nrows, FArgs...);
+
+    cudaDeviceSynchronize();
+    CUDA_CHECK_LAST();
+
+    POP_RANGE();
+  }
+
+  /**
+   * @brief If we cannot perform aggregation on this type then throw an error
+   */
+  template<typename ColumnType, template <typename AggType> class agg_op, bool average, class... TArgs,
+	   typename std::enable_if_t<!std::is_arithmetic<ColumnType>::value, std::nullptr_t> = nullptr> 
+  void dispatch_aggregation_type(gdf_size_type nrows, TArgs... FArgs)
+  {
+    CUDF_FAIL("Unsupported column type. Only arithmetic types are supported in rolling windows");
+  }
+
+  /**
+   * @brief Helper function for gdf_rolling. Deduces the type of the
+   * aggregation column and type and calls another function to invoke the
+   * rolling window kernel.
+   */
+  template <typename ColumnType>
+  void operator()(gdf_size_type nrows,
+		  gdf_agg_op agg_type,
+		  void *out_col_data_ptr, gdf_valid_type *out_col_valid_ptr,
+		  void *in_col_data_ptr, gdf_valid_type *in_col_valid_ptr,
+		  gdf_size_type window,
+		  gdf_size_type min_periods,
+		  gdf_size_type forward_window,
+		  const gdf_size_type *window_col,
+		  const gdf_size_type *min_periods_col,
+		  const gdf_size_type *forward_window_col)
+  {
+    switch (agg_type) {
+    case GDF_SUM:
+      dispatch_aggregation_type<ColumnType, sum_op, false>(nrows, reinterpret_cast<ColumnType*>(out_col_data_ptr), out_col_valid_ptr, reinterpret_cast<ColumnType*>(in_col_data_ptr), in_col_valid_ptr, window, min_periods, forward_window, window_col, min_periods_col, forward_window_col);
```

Comment:
In order to minimize repeated code and the size of these lines, I suggest doing something like:

(also note: use `static_cast` instead of `reinterpret_cast`)
```
ColumnType * typed_out_data = static_cast<ColumnType*>(out_col_data_ptr);
ColumnType const * = static_cast<ColumnType const*>(in_col_data_ptr);
...
dispatch_aggregation_type<ColumnType, sum_op, false>(nrows, typed_out_data, out_col_valid_ptr, typed_in_data, in_col_valid_ptr, window, min_periods, forward_window, window_col, min_periods_col, forward_window_col);
```